benches/r1cs.rs

```diff
@@ -1,247 +1,129 @@
-            .unzip();
-
-        ShuffleProof::gadget(&mut prover, input_vars, output_vars)?;
-
-        let proof = prover.prove(bp_gens)?;
-
-        Ok((ShuffleProof(proof), input_commitments, output_commitments))
-    }
+/// Benchmark a verifier with a given number of constraints
+fn bench_verifier_with_size(n_constraints: usize, c: &mut Criterion) {
+    assert!(n_constraints.is_power_of_two());
+
+    let mut group = c.benchmark_group("r1cs");
+    let benchmark_id = BenchmarkId::new("verifier", n_constraints);
+    group.bench_function(benchmark_id, |b| {
+        // Prove a statement
+        let (proof, _duration) = prove_sized_statement_with_timer(n_constraints);
+
+        b.iter_custom(|n_iters| {
+            let mut total_time = Duration::ZERO;
+
+            for _ in 0..n_iters {
+                // Setup the verifier
+                let pc_gens = PedersenGens::default();
+                let mut transcript = HashChainTranscript::new(b"test");
+                let mut verifier = Verifier::new(&pc_gens, &mut transcript);
+
+                let bp_gens = BulletproofGens::new(n_constraints, 1 /* party_capacity */);
+
+                // Apply the constraints
+                let mut var = verifier.commit(random_point());
+                for _ in 0..n_constraints {
+                    let (_, _, new_var) = verifier.multiply(var.into(), var.into());
+                    var = new_var;
+                }
+
+                // Verify the proof
+                let start_time = Instant::now();
+                assert!(black_box(verifier.verify(&proof, &bp_gens)).is_err());
+                total_time += start_time.elapsed();
+            }
+
+            total_time
+        });
+    });
 }
 
-impl ShuffleProof {
-    /// Attempt to verify a `ShuffleProof`.
-    pub fn verify(
-        &self,
-        pc_gens: &PedersenGens,
-        bp_gens: &BulletproofGens,
-        transcript: &mut Transcript,
-        input_commitments: &[StarkPoint],
-        output_commitments: &[StarkPoint],
-    ) -> Result<(), R1CSError> {
-        // Apply a domain separator with the shuffle parameters to the transcript
-        // XXX should this be part of the gadget?
-        let k = input_commitments.len();
-        transcript.append_message(b"dom-sep", b"ShuffleProof");
-        transcript.append_u64(b"k", k as u64);
-
-        let mut verifier = Verifier::new(pc_gens, transcript);
-
-        let input_vars: Vec<_> = input_commitments
-            .iter()
-            .map(|V| verifier.commit(*V))
-            .collect();
-
-        let output_vars: Vec<_> = output_commitments
-            .iter()
-            .map(|V| verifier.commit(*V))
-            .collect();
-
-        ShuffleProof::gadget(&mut verifier, input_vars, output_vars)?;
-
-        verifier.verify(&self.0, bp_gens)
-    }
-}
+/// Prove a sized circuit and time *only* the prover latency
+fn prove_sized_statement_with_timer(n_constraints: usize) -> (R1CSProof, Duration) {
```

Comment:
nit:

would be kinda nice to make the portion of this fn after the prover has been instantiated be a helper fn which accepts a generic `CS: RandomizedConstraintSystem` and uses `commit_public`, then this could be reused between `prove_sized_statement_with_timer` and the inner loop of `bench_verifier_with_size`.

Reply:
good suggestion, done